Added version comparison and tests

cjrolo opened the pull request:
Closes #146

This PR addresses issue #146 in the following way:

1. Adds Version information in the `CompressorHeader` (Extracted from `Cargo.toml`)
2. On Decoding checks the version of the file against the version of the compressor.
  2.1 Doesn't allow decompression of higher version files by lower version compressor (Should it?)
3. Add tests for the compression/decompression
4. Minor version bump to start tagging the files where the header was introduced.

worryg0d commented on atsc/src/header.rs:
```diff
@@ -14,19 +14,72 @@ See the License for the specific language governing permissions and
+        };
+        let current_version = env!("CARGO_PKG_VERSION").to_string();
+        trace!("Versions: c:{} h:{}", current_version, header.version);
+        match compare(current_version.clone(), header.version.clone()) {
```
Nit: Referring to `version-compare` [source code](https://gitlab.com/timvisee/version-compare/-/blob/master/src/compare.rs?ref_type=heads#L29) it looks like it is possible to avoid cloning by passing references to versions:
```suggestion
        match compare(&current_version, &header.version) {
```

worryg0d reviewed (approved):
LGTM! I left some nits regarding some code duplications and little optimization.

> 2.1 Doesn't allow decompression of higher version files by lower version compressor (Should it?)

I think so. The frame format may be changed in the future, so parsing them with an outdated compressor version may lead to undefined behavior.

cjrolo commented:
@rukai Can you give a look at this PR please. Thanks!

rukai commented:
I have two concerns:
1. I am worried about coupling the file version to the crate version.
2. Storing the version in bincode prevents us from actually reading older versions due to how the bincode protocol works.

## 1. Crate versioning vs file versioning

The needs for versioning the ATSC crate is:
* Enable changing the API of the ATSC crate which is used as a library for e.g. https://github.com/instaclustr/atsc/blob/main/csv-compressor/Cargo.toml

Semantic versioning is important here since it enables interoperability across the crate ecosystem while allowing non-breaking changes to be easily pulled in via `cargo update`.

The needs for file format versioning is:
* Allow making breaking changes to the file format

Semantic versioning is overkill here.
A simple incrementing version is all thats needed since all changes to the file format are breaking and ATSC will be required to automatically upgrade older versions to its native version.

### Concerns with using the same version

* What if we want to change the crate API without changing the file format?
* What if we want to change the file format without changing the crate API?
* During development we want to be able to bump the the file format version as file format changes occurs (possibly multiple times before a crate release occurs) so that developers can avoid breaking their test files. However the crate version is usually bumped only just before release of a new version.
* What if we move the serialization/deserialization logic into its own crate? Then we wont be able to access the ATSC version via `CARGO_PKG_VERSION` like we do currently.


### Proposed Solution

So instead I propose that we store the version of the file as a u32.
We would then keep a CURRENT_VERSION constant in ATSC that we increment any time a breaking change is made to the file format.
It would be easy enough to remember to do this since we would need to write a transform for each upgrade anyway. And these transforms should of course have corresponding tests written for them.

## 2. store the version out of bincode

Unlike protocols like json or cbor, [bincode does not provide any mechanism for performing upgrades on old file formats](https://github.com/bincode-org/bincode?tab=readme-ov-file#is-bincode-suitable-for-storage).
I also believe that having ATSC perform upgrades on old file formats is an important requirement.

So, if we are to continue using bincode then we need to store the version out of band so that we can deserialize to the old bincode format and then upgrade the deserialized values to the new format.
This approach will also enable us to move off bincode in the future if we ever had the need.

I propose we do this by storing the u32 of the version as the first 4 bytes of the file and then begin decoding the bincode after reading those 4 bytes.

rukai commented:
I took a look at https://github.com/instaclustr/atsc/issues/146 and can see how you ended up with this solution. The focus of the issue was on compatibility with the logic of the compressor rather than changes to the file format itself.

I still think that the approach I wrote up is worth considering since it enables changes to the file format as well.
If the logic of the compressor changes such that old files no longer decompress in a reasonable way then the file version should be bumped. (I wouldnt consider slight fluctuations in output values within margin of error to be worth a version bump)

But if the ability to change the file format is not important then we can disregard it.

cjrolo commented:
Amazing feedback, thanks!

I think it makes all the sense to allow header inspection before the bincode. I also think that the version as u32 is a great plan. Then then compressors don't need to tie to the compressor version, but to the file version, makes much more sense.

Going to rewrite this following this approach.

cjrolo commented:
Second pass at this!

So, header changed and includes version as a `u32`.

Also,

- Removed bincode from header: We can decompress the header (9 Bytes)
- Header version is stored in Bytes 4 to 7.
- Added additional tests and refactor part of the tests.

Input appreciated!

rukai commented on atsc/Cargo.toml:
```diff
@@ -24,6 +24,7 @@ inverse_distance_weight = "0.1.1"
 num-traits = "0.2"
 csv = "1.3.1"
 thiserror = "2.0.3"
+version-compare = "0.1"
```
This is unused and can be removed

rukai commented on atsc/src/header.rs:
```diff
@@ -14,25 +14,106 @@ See the License for the specific language governing permissions and
 
-#[derive(Encode, Decode, Debug, Clone)]
+/*  The current file version.
+    On file read, compressors check the version and uncompress accordingly to that (of fail)
```
maybe `of fail` is mean to be `or fail`?

rukai commented on atsc/src/header.rs:
```diff
@@ -14,25 +14,106 @@ See the License for the specific language governing permissions and
+    trace!("Versions: c:{} h:{}", current_version, version);
+    match current_version.cmp(&version) {
+        std::cmp::Ordering::Less => panic!(
+            "Can't decompress! File is version ({}) is higher than compressor version ({})!",
```
Grammar nit: `File is version ({}) which is higher`

rukai commented on atsc/src/header.rs:
```diff
@@ -14,25 +14,106 @@ See the License for the specific language governing permissions and
+        // Extract initial_segment
+        let initial_segment = [data[0], data[1], data[2], data[3]];
+        // Extract version (u32 from 4 bytes)
+        let version = u32::from_le_bytes([data[4], data[5], data[6], data[7]]);
```
Since the version will always be the first part of the header to be parsed, I think the version should be the first 4 bytes starting at index 0 rather than starting at index 4.

rukai commented on atsc/src/header.rs:
```diff
@@ -14,25 +14,106 @@ See the License for the specific language governing permissions and
             initial_segment: *b"BRRO",
-            // We have to limit the bytes of the header
+            version: CURRENT_VERSION,
             frame_count: 0,
```
Technically this always existed, but previously we werent actually writing this to the file.
Why does this field exist? Cant we just call `self.data_frames.len()` to get the number of frames?

cjrolo replied:
I'm trying to future proof in case you want to decode a specific frame. You could read the header, take the number of frames and go from there without decompressing the whole file. Right now, it doesn't bring additional functionality/utility.

rukai replied:
I dont think thats actually possible with bincode.
If you wanted to go that route I think the best way would be to manually manage the encoding/decoding of frames and have bincode encode/decode at the individual frame level.

But thats definitely out of scope of this PR.
I'll leave you to decide the direction you want to take this.

cjrolo replied:
You just described everything I think needs to be done. The manually manage encoding/decoding of frames would be the next step. I feel the current bincode approach, while it was fast, it's causing some limitations on improving decompression capacities. I will open a PR/Discussion thread to get this better defined.

rukai commented on atsc/src/header.rs:
```diff
@@ -14,25 +14,106 @@ See the License for the specific language governing permissions and
+        self.frame_count
+    }
+
+    pub fn to_bytes(&self) -> Vec<u8> {
```
To reduce allocations I would define this as:
`pub fn to_bytes(&self, writer: &mut Vec<u8>)`
And then write to the passed in vec instead of creating a new one.

rukai commented on atsc/src/header.rs:
```diff
@@ -14,25 +14,106 @@ See the License for the specific language governing permissions and
 impl CompressorHeader {
     pub fn new() -> Self {
         CompressorHeader {
             initial_segment: *b"BRRO",
```
Personally I would just write the BRRO bytes directly in `to_bytes` and return an Err in `from_bytes` if the magic bytes are not BRRO.
But this is a very small nit either approach is fine.

rukai commented on atsc/src/data.rs:
```diff
@@ -70,21 +72,34 @@ impl CompressedStream {
         let config = BinConfig::get();
-        bincode::encode_to_vec(self, config).unwrap()
+        out.extend_from_slice(&self.header.to_bytes());
+        out.extend_from_slice(&bincode::encode_to_vec(self.data_frames, config).unwrap());
```
this line will write the body of the file (99% of the file) into one buffer and then copy the result of that into another buffer (`out`) and most likely trigger an allocation in the process.

To avoid this, I think we can instead write the line as `bincode::encode_into_std_write(self.data_frames, &mut out, config).unwrap();`.
I havent tried compiling this though.

rukai reviewed (approved):
Looks good!
